go-controller/pkg/libovsdbops/dhcp.go

```diff
@@ -0,0 +1,71 @@
+package libovsdbops
+
+import (
+	"reflect"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+type dhcpOptionsPredicate func(*nbdb.DHCPOptions) bool
+
+func CreateOrUpdateDhcpOptionsOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, lsp *nbdb.LogicalSwitchPort, dhcpv4Options, dhcpv6Options *nbdb.DHCPOptions) ([]libovsdb.Operation, error) {
+	opModels := []operationModel{}
+	if dhcpv4Options != nil {
+		opModels = append(opModels, operationModel{
+			Model: dhcpv4Options,
+			ModelPredicate: func(item *nbdb.DHCPOptions) bool {
+				return item.Cidr == dhcpv4Options.Cidr && reflect.DeepEqual(item.ExternalIDs, dhcpv4Options.ExternalIDs)
+			},
```

Comment:
So I wonder about two things here:

- Why are we not checking thew options column in the predicate?
- Do we want the same dhcp options row to be potentially associated with multiple LSPs as long as they have the same CIDR and external IDs? Or would it be not possible for that to happen given how we are making use of it?

Reply:
Consider generating a unique Id for each DHCPOptions (the LSP id) as an external ID and only match on that.

See pkg/libovsdbops/db_object_ids.go

Reply:
For IPv4 DHCP Uniqueness is per VM name + cidr since everything else is always the same, for IPv6 we cannot pass hostname since it's not implemented at OVN so it's just per cidr.

Reply:
> External IDs + CIDR is enough for the kind of uniqueness that we need.

Consider you are working on a generic package that might need to cater for future needs. You are hardcoding a predicate within the package that might not be an absolute truth forever. We have gotten in trouble in the past because of this so since then we have a tendency to pass the predicate in. So consider that if possible. But not a hard ask, I guess it is fine as it is as long as we don't have future uses for DHCP options.

> 
> > * Do we want the same dhcp options row to be potentially associated with multiple LSPs as long as they have the same CIDR and external IDs? Or would it be not possible for that to happen given how we are making use of it?
> 
> Yes, since at live migration we will have two LSPs with same DHCP Options (same cidr same vm name)

There is also the option to dup the DHCP options I guess? Although less optimal, agreed.
> 
> > Consider generating a unique Id for each DHCPOptions (the LSP id) as an external ID and only match on that.
> > See pkg/libovsdbops/db_object_ids.go
> 
> But this will generate a DHCPOptions per LSP and we don't need that

Not necessarily. Perhaps for your use case you can define an ID that is just based on the CIDR? Could the CIDR be used for uniqueness in your use case?

Using an ID instead of a predicate allows us to potentially index it in the client in case we need it. If we need it in the future but we don't do it now, we could end up needing  a DB migration process.

Reply:
> Not necessarily. Perhaps for your use case you can define an ID that is just based on the CIDR? Could the CIDR be used for uniqueness in your use case?

I can define and ID with hostname + CIDR for IPv4 and just CIDR for IPv6.

> 
> Using an ID instead of a predicate allows us to potentially index it in the client in case we need it. If we need it in the future but we don't do it now, we could end up needing a DB migration process.

I will try to generate one so we don't harcode the predicate at the function